miden/src/cli/data.rs

```diff
@@ -390,10 +391,15 @@ pub struct ProgramFile {
 /// Helper methods to interact with masm program file.
 impl ProgramFile {
     /// Reads the masm file at the specified path and parses it into a [ProgramFile].
-    #[instrument(name = "read_program_file", fields(path = %path.display()))]
-    pub fn read(path: &PathBuf) -> Result<Self, Report> {
+    #[instrument(name = "read_program_file", skip(source_manager), fields(path = %path.display()))]
+    pub fn read(
+        path: &PathBuf,
+        source_manager: &dyn assembly::SourceManager,
+    ) -> Result<Self, Report> {
```

Comment:
Same situation as above - we need to re-use the `SourceManager` so that source spans remain valid throughout the lifetime of the sources we parsed with it. If we use a new source manager, then it won't cause bugs necessarily, but it will result in being unable to render diagnostics correctly.

Reply:
Makes sense re needing to re-use the `SourceManager`. Though, here I was thinking that we could re-use it by putting it into the struct. Basically:

- In this method we instantiate a new `SingleThreadedSourceManger` which we then use to parse the file.
- This source manager is "saved" in the `ProgramFile` struct.
- This way, when we call `ProgramFile::compile()` we can re-use the same source manager.

The main benefit that I see here is that we only need to change `SourceFile` struct and the users of this struct don't need to know about source managers.

Reply:
The problem is that we aren't using this just with one file (the program), the program itself is just one module, but `compile` also accepts a set of libraries, and we need the source manager to know about all of those sources as well, in order to render source-level debug information during program execution, and those libraries may have been compiled on the fly as well.

That said, I think a solution that might address your point here is to require passing in an `Arc<dyn SourceManager>` to `read`, and store that in the resulting `ProgramFile` struct, so that `compile` doesn't need to take it as an argument as well. This is one of the reasons why `SourceManager` is designed the way it is (so that it can support storing a reference to it in structs like this).

Reply:
> The problem is that we aren't using this just with one file (the program), the program itself is just one module, but `compile` also accepts a set of libraries,

Looking through the code, I couldn't find a place where we use it for compiling something more than just a single executable module (i.e., we create a source manager, instantiate a `ProgramFile` and then use the source manager in `ProgramFile::compile()` - but never use it after that).

Unless I've missed something, I think my suggestion should work.